source/server/listener_impl.h

```diff
@@ -74,12 +76,130 @@ class ListenSocketFactoryImpl : public Network::ListenSocketFactory,
+ * listeners share the same ListenerFactoryContextBaseImpl. With ListenerFactoryContextBase, the
+ * number of listener config at runtime is restricted to 1, despite the active filter chains could
+ * spread among multiple listener configs.
+ */
+class ListenerFactoryContextBaseImpl final : public Configuration::FactoryContext,
+                                             public Network::DrainDecision {
+public:
+  ListenerFactoryContextBaseImpl(Envoy::Server::Instance& server,
+                                 ProtobufMessage::ValidationVisitor& validation_visitor,
+                                 const envoy::config::listener::v3::Listener& config,
+                                 Server::DrainManagerPtr drain_manager);
+  AccessLog::AccessLogManager& accessLogManager() override;
+  Upstream::ClusterManager& clusterManager() override;
+  Event::Dispatcher& dispatcher() override;
+  Network::DrainDecision& drainDecision() override;
+  Grpc::Context& grpcContext() override;
+  bool healthCheckFailed() override;
+  Http::Context& httpContext() override;
+  Init::Manager& initManager() override;
+  const LocalInfo::LocalInfo& localInfo() const override;
+  Envoy::Runtime::RandomGenerator& random() override;
+  Envoy::Runtime::Loader& runtime() override;
+  Stats::Scope& scope() override;
+  Singleton::Manager& singletonManager() override;
+  OverloadManager& overloadManager() override;
+  ThreadLocal::Instance& threadLocal() override;
+  Admin& admin() override;
+  const envoy::config::core::v3::Metadata& listenerMetadata() const override;
+  envoy::config::core::v3::TrafficDirection direction() const override;
+  TimeSource& timeSource() override;
+  ProtobufMessage::ValidationContext& messageValidationContext() override;
+  ProtobufMessage::ValidationVisitor& messageValidationVisitor() override;
+  Api::Api& api() override;
+  ServerLifecycleNotifier& lifecycleNotifier() override;
+  ProcessContextOptRef processContext() override;
+  Configuration::ServerFactoryContext& getServerFactoryContext() const override;
+  Configuration::TransportSocketFactoryContext& getTransportSocketFactoryContext() const override;
+  Stats::Scope& listenerScope() override;
+
+  // DrainDecision
+  bool drainClose() const override {
+    return drain_manager_->drainClose() || server_.drainManager().drainClose();
+  }
+  Server::DrainManager& drainManager();
+
+private:
+  Envoy::Server::Instance& server_;
+  const envoy::config::core::v3::Metadata metadata_;
+  envoy::config::core::v3::TrafficDirection direction_;
+  Stats::ScopePtr global_scope_;
+  Stats::ScopePtr listener_scope_; // Stats with listener named scope.
+  ProtobufMessage::ValidationVisitor& validation_visitor_;
+  const Server::DrainManagerPtr drain_manager_;
+};
+
+class ListenerImpl;
+// TODO(lambdai): Strip the interface since ListenerFactoryContext only need to support
+// ListenerFilterChain creation.
```

Comment:
Please do this now. It's very hard to understand the new flow of classes/interfaces here. If you need to do an intermediate PR let's do that.

Reply:
@mattklein123  Sorry I propose delay this strip. Previously ListenerFactoryContext supports both network filter chain creation and listener filter chain creation. 
So ListenerFactoryContext must be fat enough to cover both.
Obviously when we split the creation the two interface could be smaller but it take time to see if the filters in envoy repo and in 3rd party repo could survive. I assume we need a grace period to announce (even simply through issues) and to deprecate.

Reply:
> Obviously when we split the creation the two interface could be smaller but it take time to see if the filters in envoy repo and in 3rd party repo could survive. I assume we need a grace period to announce (even simply through issues) and to deprecate.

We don't promise API compatibility so this can be done now, or please make it more clear in the TODO what needs to be done. The comment is not thorough enough to understand.